### PR TITLE
fix for percentage dimensions

### PR DESCRIPTION
**Pull request by willmcgugan**
Calculating percentage dimensions didn't take in to account padding / border.

Fixes https://github.com/Textualize/textual/issues/3721

Supersedes https://github.com/Textualize/textual/pull/4023

### CODE REVIEW

File: src/textual/widget.py [hunk 1]
```diff
@@ -1143,25 +1137,21 @@ def _get_box_model(
             styles_height = styles.height
             # Explicit height set
             content_height = styles_height.resolve(
-                sizing_container - styles.margin.totals, viewport, height_fraction
+                container - styles.margin.totals, viewport, height_fraction
             )
-            if is_border_box and styles_height.excludes_border:
```
Comment (willmcgugan) [inline src/textual/widget.py:1148]: Not sure what I was thinking re `excludes_border`. But it doesn't appear to be required.

**Comment by rodrigogiraoserrao** [comment]
There's something wrong with this PR.

If you look at `dimensions_width.py` test in the superseded PR, you'll see that your PR doesn't pass that test.
Barring any misunderstanding on my part of how `max-`/`min-` units are supposed to work, the apps from `dimensions_width.py`, `dimensions_max_width.py`, and `dimensions_min_width.py`, all should look the same but they don't (I suggest looking at them on a terminal of width 80 because that's the length of the ruler).

**Comment by willmcgugan** [comment]
@rodrigogiraoserrao Could you please isolate a single case that isn't working. i.e. an MRE with as few widgets as possible which displays incorrect behaviour.

**Comment by rodrigogiraoserrao** [comment]
Sure, here is one: `%` isn't respected when `margin` is set.
In this case, the red frame should take up all of the horizontal width but it's off by 4

```py
from textual.app import App, ComposeResult
from textual.widgets import Label


class NestedCSSTokenErrorApp(App[None]):
    CSS = """
    Label.target {
        width: 100%;
        margin: 2;
        border: solid red;
    }
    """

    def compose(self) -> ComposeResult:
        # yield Label("1234567890" * 80)
        yield Label("", classes="target")


if __name__ == "__main__":
    NestedCSSTokenErrorApp().run()
```

Another manifestation of the same problem can be seen in the app below. The bottom label is correctly set to 50% width (via `max-width`) but the top one isn't:

```py
from textual.app import App, ComposeResult
from textual.widgets import Label, Static


class NestedCSSTokenErrorApp(App[None]):
    CSS = """
    Label {
        border: solid red;
        margin: 0 4;
    }
    #first { width: 50%; }
    #second {
        width: 100%;
        max-width: 50%;
    }
    """

    def compose(self) -> ComposeResult:
        yield Static("1234567890" * 8)
        yield Label("", id="first")
        yield Label("", id="second")


if __name__ == "__main__":
    NestedCSSTokenErrorApp().run()
```

**Comment by rodrigogiraoserrao** [comment]
This may be why you feel my PR changed the semantics of `%`, but we talked and we agreed that the margin is _not_ a part of the size of the widget.

When we talked, you mentioned there was some special case in the code to do something cheeky with `100%`, widgets with a single children, and margin, but we also realised that you didn't need that special case (which impacted other situations where there are % units and margin, like the one with the `50%` above).

**Comment by willmcgugan** [comment]
It's behaving as I'd expect.

The percentage applies to the available width, which doesn't include margin.

If it didn't behave like this, and you set width 100% with a margin, it would be pushed off screen. It would be simpler (for us) if the percentage was an exact percentage of the container, but this would make it very hard for the dev to achieve sensible layouts.

The behaviour should be identical if you have a width of `1fr` with a single item.

**Comment by rodrigogiraoserrao** [comment]
I was under the impression that the percent unit `%` was relative to the _total size of the container_ and not _“available” size of the container_, and that using margins that are too big with sizes that are too big does result in things getting pushed off.

I know we don't blindly copy browser CSS but it appears that's what it does as well:

Screenshot 2024-01-17 at 17 40 05

When we talked, we understood that the effect you're describing could be achieved by setting padding in the container and child dimensions to `fr` units, for example, precisely because `100%` + `margin` pushes things off screen.

**Comment by TomJGooding** [comment]
Hope you don't mind me sticking my nose in as usual, but this behaviour with margin seems counter-intuitive to me?

**Percent size with padding :white_check_mark:**

<details>
<summary>Code</summary>

```python3
from textual.app import App, ComposeResult
from textual.containers import Container
from textual.widgets import Static


class ExampleApp(App):
    CSS = """
    Screen {
        align: center middle;
    }

    Container {
        height: 10;
        width: 10;
        background: blue;
    }

    Static {
        height: 50%;
        width: 50%;
        background: red;
        padding: 2 1;
    }
    """

    def compose(self) -> ComposeResult:
        with Container():
            yield Static("5x5")


if __name__ == "__main__":
    app = ExampleApp()
    app.run()
```

</details>

image


**Percent size with padding and margin :x:**

<details>
<summary>Code</summary>

```python3
from textual.app import App, ComposeResult
from textual.containers import Container
from textual.widgets import Static


class ExampleApp(App):
    CSS = """
    Screen {
        align: center middle;
    }

    Container {
        height: 10;
        width: 10;
        background: blue;
    }

    Static {
        height: 50%;
        width: 50%;
        background: red;
        padding: 2 1;
        margin-top: 5;
        margin-left: 5;
    }
    """

    def compose(self) -> ComposeResult:
        with Container():
            yield Static("5x5")


if __name__ == "__main__":
    app = ExampleApp()
    app.run()
```

</details>

image

**Comment by willmcgugan** [comment]
It's a pragmatic design choice for Textual. If I have a widget with width and height 100% and apply a margin, it currently produces something useful. i.e. the widget is centered in the screen. If the margin was additive, then the widget would overflow the screen boundary. That is very unlikely to be useful for the developer.

If you do want a widget to be an exact percentage, but offset from the top left then you can use `offset` and achieve the same results.

The interaction with margin is orthogonal to the original issue. The original issue was that percentage dimensions excluded the gutter, which this PR should fix. Nothing re margins has changed. We can always change that if needed, but AFAIK nobody as found this to be a problem, and I think we should err on the side of not breaking things.

File: src/textual/layouts/horizontal.py [hunk 2]
```diff
@@ -26,7 +26,9 @@ def arrange(
         add_placement = placements.append
 
         child_styles = [child.styles for child in children]
-        box_margins: list[Spacing] = [styles.margin for styles in child_styles]
+        box_margins: list[Spacing] = [
+            styles.margin for styles in child_styles if styles.overlay != "screen"
+        ]
```
Comment (rodrigogiraoserrao) [inline src/textual/layouts/horizontal.py:31]: I see you changed this because you probably thought of an edge case that wasn't being covered; wouldn't this warrant a test?

File: src/textual/layouts/horizontal.py [hunk 3]
```diff
@@ -36,7 +38,7 @@ def arrange(
                     ]
                 )
                 + (box_margins[0].left + box_margins[-1].right),
-                max(
+                min(
```
Comment (rodrigogiraoserrao) [inline src/textual/layouts/horizontal.py:41]: Same comment re: test

File: src/textual/layouts/vertical.py [hunk 4]
```diff
@@ -29,7 +29,7 @@ def arrange(
         ]
         if box_margins:
             resolve_margin = Size(
-                max(
+                min(
```
Comment (rodrigogiraoserrao) [inline src/textual/layouts/vertical.py:32]: Same comment re: test

File: src/textual/widget.py (hunk 1 above)
Comment (rodrigogiraoserrao) [reply src/textual/widget.py:1148]: I agree it doesn't appear to be required at all.

**Review by rodrigogiraoserrao — COMMENTED** [review]
If `margin` + `%` is to work like so, shouldn't `margin` be taken into account for `max`/`min` units, then?

E.g., regardless of the way we decide to resolve `%` units with margin, I'd expect the two placeholders below to have the same width:
Screenshot 2024-01-18 at 14 22 09

<details>
<summary>App code:</summary>

```py
from textual.app import App, ComposeResult
from textual.widgets import Label, Placeholder


class NestedCSSTokenErrorApp(App[None]):
    CSS = """
    Placeholder {
        margin: 0 4;
        height: 3;
    }
    #first { 
        width: 50%;
    }
    #second {
        width: 999;
        max-width: 50%;
    }
    """

    def compose(self) -> ComposeResult:
        yield Label("1234567890" * 8)
        yield Placeholder("", id="first")
        yield Placeholder("", id="second")


if __name__ == "__main__":
    NestedCSSTokenErrorApp().run()
```

</details>

**Comment by TomJGooding** [comment]
> It's a pragmatic design choice for Textual. If I have a widget with width and height 100% and apply a margin, it currently produces something useful. i.e. the widget is centered in the screen.

I promise this will be my last post on this PR, but surely this is where you should add padding to the container? I'm afraid I still don't understand the reason for changing the size of a widget based on its margin.

File: src/textual/layouts/horizontal.py (hunk 2 above)
Comment (willmcgugan) [reply src/textual/layouts/horizontal.py:31]: It probably should. Going to leave it for now, because overlay isn't yet documented.

File: src/textual/layouts/horizontal.py (hunk 3 above)
Comment (willmcgugan) [reply src/textual/layouts/horizontal.py:41]: It is covered by a snapshot, but it probably does deserve a more specific test. Going to leave it for now though.

**Comment by rodrigogiraoserrao** [comment]
@willmcgugan you didn't say anything regarding [this comment](https://github.com/Textualize/textual/pull/4037#pullrequestreview-1829781345) so I'm not sure if this slipped past you or if it's the intended behaviour.
